lib/puppet/provider/vault_cert/openssl.rb

```diff
@@ -0,0 +1,308 @@
+
+    # Verify that a response was returned
+    if response.body
+      response_body = JSON.parse(response.body)
+      # Check the revocation time on the cert object
+      if response_body['data']['revocation_time'] > 0
+        return true
+      else
+        return false
+      end
+    else
+      return true
+    end
+  end
+
+  # Save an openssl cert object into the global cert var
+  def get_certificate
+    return @cert unless @cert.nil?
+    @cert = if Pathname.new(resource[:cert_path]).exist?
+              file = File.read(resource[:cert_path])
+              OpenSSL::X509::Certificate.new(file)
+            else
+              false
+            end
+  end
+
+  # Save an openssl PKey object into the global priv_key var
+  def get_private_key
+    return @priv_key unless @priv_key.nil?
+    @priv_key = if Pathname.new(resource[:priv_key_path]).exist?
+                  file = File.read(resource[:priv_key_path])
+                  if resource[:auth_type] == 'dsa'
+                    #OpenSSL::PKey::DSA.new(file, resource[:password])
+                    OpenSSL::PKey::DSA.new(file)
+                  elsif resource[:auth_type] == 'rsa'
+                    #OpenSSL::PKey::RSA.new(file, resource[:password])
+                    OpenSSL::PKey::RSA.new(file)
+                  elsif resource[:auth_type] == 'ec'
+                    #OpenSSL::PKey::EC.new(file, resource[:password])
+                    OpenSSL::PKey::EC.new(file)
+                  else
+                     raise Puppet::Error,
+                           "Unknown authentication type '#{resource[:auth_type]}'"
+                  end
+                  #OpenSSL::PKey::RSA.new(file)
+                else
+                  false
+                end
+  end
+
+  # Read the serial number from the certificate, convert it to base 16, and add colons
+  def get_cert_serial
+    cert = get_certificate
+    # Get the serial number from the cert object
+    serial_number = cert.serial.to_s(16)
+    # Add a colon every 2 characters to the returned serial number
+    serial_number.scan(/\w{2}/).join(':')
+  end
+
+  # Save the certificate and private key on the client server
+  def client_cert_save(cert)
+    cert_name = get_client_fqdn + '.crt'
+    # Get the directory from the given cert path
+    cert_dir = File.dirname(resource[:cert_path])
+    # Save the new cert in the certs directory on the client server
+    cert_path = File.join(cert_dir, cert_name)
+    File.open(cert_path, 'w') do |f|
+      f.write(cert['data']['certificate'])
+    end
+
+    key_name = get_client_fqdn + '.key'
+    # Get the directory from the given key path
+    key_dir = File.dirname(resource[:priv_key_path])
+    # Save the new private key in the tls directory on the client
+    key_path = File.join(key_dir, key_name)
+    File.open(key_path, 'w') do |f|
+      f.write(cert['data']['private_key'])
+    end
+  end
+
+  # TODO: Save a copy of the newly generated cert on the vault server
```

Comment:
is this TODO still valid?

Reply:
No , I'll remove it since it isn't necessary